### PR TITLE
[FIXED] JetStream flow control may stall in some conditions

### PR DESCRIPTION
**Pull request by kozlovic**
Keep track of the "sequence" of incoming messages that are then compared with the number of delivered message.
However, ChanSubscription were broken: the FC would *always* be responded to when received, because pMsgs (and pBytes) would always be 0 for these type of subscriptions.

We now check for flow control on all incoming messages for ChanSubscription (and evaluate the "delivered" count based on the incoming sequence minus length of the user's channel). We also have a long-lived go routine to cover situations where the evaluation during the incoming message did not result in the target to send the FC, but then the user consumes enough messages to bring the "delivered" at or past the target. Without this go routine, the client would stall and have to wait for an heartbeat with the "fc stalled" header in order to be unblocked.

Alternatively, we could fail creation of ChanSubscription subscriptions if flow control is in use... 

Finally, this PR fixes some issues with the ordered consumer test:
- We send a "EOF" empty message which should not be counted as the number of chunks used to reconstitute the asset
- Some "message filters" that are removed as part of the execution of the filter's callback would not be put back for the following "sync" test (we test async then sync).

Signed-off-by: Ivan Kozlovic <ivan@synadia.com>

### CODE REVIEW

File: js.go [hunk 1]
```diff
@@ -1581,7 +1636,7 @@ func (sub *Subscription) resetOrderedConsumer(sseq uint64) {
 		// Reset some items in jsi.
 		jsi.dseq = 1
 		jsi.cmeta = _EMPTY_
-		jsi.fcr, jsi.fcd = _EMPTY_, 0
+		jsi.fcr, jsi.fcd, jsi.fciseq = _EMPTY_, 0, 0
```
Comment (kozlovic) [inline js.go:1639]: I just made that change (with push force) after submitting the PR but I think it is wrong, since the subscription is maintained, the sub.delivered count is not reset either, so this value should not be reset. Will revert this change and push force again.

File: js.go [hunk 2]
```diff
@@ -1442,10 +1463,41 @@ func (js *js) subscribe(subj, queue string, cb MsgHandler, ch chan *Msg, isSync,
 	if hasHeartbeats {
 		sub.scheduleHeartbeatCheck()
 	}
+	// For ChanSubscriptions, if we know that there is flow control, we will
+	// start a go routine that evaluates the number of delivered messages
+	// and process flow control.
+	if sub.Type() == ChanSubscription && hasFC {
+		go sub.chanSubcheckForFlowControlResponse()
```
Comment (derekcollison) [inline js.go:1470]: Why does it have to be in Go routine?

File: js.go [hunk 3]
```diff
@@ -1442,10 +1463,41 @@ func (js *js) subscribe(subj, queue string, cb MsgHandler, ch chan *Msg, isSync,
...
+// This long-lived routine is used per ChanSubscription to check
+// on the number of delivered messages and check for flow control response.
+func (sub *Subscription) chanSubcheckForFlowControlResponse() {
+	sub.mu.Lock()
+	nc := sub.conn
+	sub.mu.Unlock()
+	if nc == nil {
+		return
+	}
+	t := time.NewTicker(chanSubFCCheckInterval)
```
Comment (derekcollison) [inline js.go:1485]: Should probably have a chan here that is closed when sub is unsubscribed or connection is closed.

File: js.go [hunk 4]
```diff
@@ -1442,10 +1463,41 @@ func (js *js) subscribe(subj, queue string, cb MsgHandler, ch chan *Msg, isSync,
...
+	for range t.C {
+		sub.mu.Lock()
+		if sub.closed {
+			sub.mu.Unlock()
+			t.Stop()
```
Comment (derekcollison) [inline js.go:1490]: Just do top level defer for when you leave the function.

File: js.go [hunk 5]
```diff
@@ -1442,10 +1463,41 @@ func (js *js) subscribe(subj, queue string, cb MsgHandler, ch chan *Msg, isSync,
...
+			return
+		}
+		fcReply := sub.checkForFlowControlResponse()
+		sub.mu.Unlock()
+		if fcReply != _EMPTY_ {
```
Comment (derekcollison) [inline js.go:1495]: If we don't check for errors just send blindly? No need to check fcReply?

File: js.go (hunk 2 above)
Comment (kozlovic) [reply js.go:1470]: Doesn't have to be... could add a timer.

File: js.go (hunk 3 above)
Comment (kozlovic) [reply js.go:1485]: Could add a chan too (in addition to the timer). I thought that it was ok like this, meaning that this go routine may linger for 250ms and then we would detect that the subscription is gone.

File: js.go (hunk 4 above)
Comment (kozlovic) [reply js.go:1490]: Have to check for each tick if the subscription is closed, so will need to get the lock here anyway.

File: js.go (hunk 5 above)
Comment (kozlovic) [reply js.go:1495]: Well, sure, Conn.Publish() checks for empty subject and returns an error, but I thought that it makes more readable that we know that fcReply will be non empty only when we need to reply back to FC.

File: js.go (hunk 2 above)
Comment (kozlovic) [reply js.go]: Changed to use a timer instead.

File: js.go (hunk 4 above)
Comment (kozlovic) [reply js.go]: Refactored that as a timer's func

File: js.go (hunk 3 above)
Comment (kozlovic) [reply js.go]: Since I use timer now, it is stopped when the subscription is stopped/removed.

**Review by derekcollison — APPROVED** [review]
LGTM
